server/lib/util.js

```diff
@@ -118,3 +119,19 @@ function dateFormatter(rows, dateFormat) {
 
   return rows;
 }
+
+// this function separate amounts by curreny_id
+
+function sumAmounts(rows) {
```

Comment:
Could you do this in SQL?  Javascript has problems calculation math correctly.

For reference, see the following issues: https://github.com/Vanga-Hospital/bhima-2.X/issues/204, https://github.com/IMA-WorldHealth/bhima-2.X/issues/1145.

For other reports that do totaling in SQL instead of JS for this reason, see [the cash report](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/controllers/finance/reports/cash/index.js#L164) and the [invoice report](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/controllers/finance/reports/invoices/index.js#L65).

For more thoughts on JS being bad at math, [here is an article on the subject](https://gooroo.io/GoorooTHINK/Article/16305/Is-Math-Broken-in-JavaScript-Part-1/18866)/

Reply:
You're right @jniles, but we have already a json array with all information we need for summing, doing another sql query won't be easy because it shoul take care all fiters or options that the previous query had.
When seaching the solution, I come across a libary witch can help to solve this probem, alasql
I'm investigating to see if good

Reply:
Just use [FilterParser](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/lib/filter.js).  It is a library @sfount wrote for this purpose.  [Here's a pretty complex example of how to use it](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/controllers/finance/journal/index.js#L81). 

In this case, you'll just need to do this:
```js
const aggregateQuery = `
SELECT c.id as currencyId, c.symbol as currencySymbol, SUM(v.amount) as totalAmount, COUNT(c.symbol) AS numVouchers
FROM voucher JOIN currency ON voucher.currency_id = currency.id
GROUP BY currency.id`;

// options here is req.query
const filters = new FilterParser(options);
  filters.period('period', 'date');
  filters.dateFrom('custom_period_start', 'date');
  filters.dateTo('custom_period_end', 'date');
  filters.equals('user_id');
  filters.equals('edited');

  filters.custom('reference', referenceStatement);

  filters.fullText('description');

  // @todo - could this be improved
  filters.custom('entity_uuid', 'v.uuid IN (SELECT DISTINCT voucher_uuid FROM voucher_item WHERE entity_uuid = ?)');

  filters.custom('type_ids', 'v.type_id IN (?)', [typeIds]);

  // @todo - could this be improved
  filters.custom('account_id', 'v.uuid IN (SELECT DISTINCT voucher_uuid FROM voucher_item WHERE account_id = ?)');

  filters.custom('invoice_uuid', REFERENCE_SQL, [options.invoice_uuid, options.invoice_uuid]);
  filters.custom('cash_uuid', REFERENCE_SQL, [options.cash_uuid, options.cash_uuid]);

  // @TODO Support ordering query (reference support for limit)?
  filters.setOrder('ORDER BY v.date DESC');

  const query = filters.applyQuery(aggregateQuery);
  const parameters = filters.parameters();

  return db.exec(aggregateQuery, parameters);
```

Since all this code already exists in [the voucher controller](https://github.com/jeremielodi/bhima-2.X/blob/7fe5c121ec16a1d17465a7bcf224e3fc9b273fda/server/controllers/finance/vouchers.js), I would suggest you make a function that simply takes in the base SQL and the options.  This function would be called `buildVoucherQuery`.  It would then return the query and parameters after applying all the filters to it.

Like this:
```js
const aggregateQuery = `
SELECT c.id as currencyId, c.symbol as currencySymbol, SUM(v.amount) as totalAmount, COUNT(c.symbol) AS numVouchers
FROM voucher JOIN currency ON voucher.currency_id = currency.id
GROUP BY currency.id`;

const { sql, params } = vouchers.buildVoucherQuery(aggregateQuery, options);
return db.exec(sql, params);
```